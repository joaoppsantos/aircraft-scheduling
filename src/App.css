.app {
	text-align: center;
}

.nav-date {
	background-color: #282c34;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	font-size: calc(10px + 2vmin);
	color: white;
	padding: 20px 0px;
}

.main-container {
	padding-top: 20px;
	font-size: 25px;
	justify-content: center;
	display: flex;
}

.schedule-container {
	width: 96vh;
	justify-content: center;
	display: flex;
	margin: 0 auto;
}

.element-container {
	margin-bottom: 20px;
}

@media screen and (max-width: 670px) {
	.main-container {
		padding-top: 20px;
		font-size: 25px;
		justify-content: center;
		display: flex;
		flex-direction: column;
	}
}