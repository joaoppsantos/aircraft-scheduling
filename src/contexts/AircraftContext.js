import { createContext, useEffect, useState } from 'react';
import request from '../request';

export const AircraftContext = createContext({});

export function AircraftProvider({ children }) {
	const [aircrafts, setAircrafts] = useState([]);
	const [isAircraftLoading, setIsAircraftLoading] = useState(true);
	const [currentAircraft, setCurrentAircraft] = useState(undefined);
	const [currentPercentage, setCurrentPercentage] = useState(0);

	const setPercentage = percentage => {
		setCurrentPercentage(percentage);
	};
	const selectAircraft = aircraft => {
		setCurrentAircraft(aircraft);
	};

	useEffect(() => {
		request.get(`/aircrafts?limit=5&offset=0`).then((response) => {
			setAircrafts(response.data.data);
			setIsAircraftLoading(false);
		});
	}, []);

	return (
		<AircraftContext.Provider value={{
			aircrafts,
			isAircraftLoading,
			currentAircraft,
			selectAircraft,
			currentPercentage,
			setPercentage
		}}>
			{children}
		</AircraftContext.Provider>
	)
}
