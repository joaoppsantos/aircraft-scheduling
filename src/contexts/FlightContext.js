import request from '../request';
import { createContext, useEffect, useState } from 'react';

export function FlightProvider({ children }) {
	const [flights, setFlights] = useState([]);
	const [activeFlights, setActiveFlights] = useState([]);
	const [offset, setOffset] = useState(-1);
	const [hasFlightsRemaining, setHasFlightsRemaining] = useState(true);
	const paginationLimit = 25;

	const toggleActiveFlight = (flight) => {
		let isFlightOnList = false;

		activeFlights.forEach((activeInfo) => {
			if (activeInfo.id === flight.id) {
				isFlightOnList = true;
			}
		});

		isFlightOnList ? 
			setActiveFlights(activeFlights.filter(activeInfo => activeInfo.id !== flight.id))
			: setActiveFlights([...activeFlights, flight]);
	}

	useEffect(() => {
		if (!hasFlightsRemaining) {
			return;
		}

		async function fetchData() {
			try {
				const newOffset = offset + 1;
				const response = await request.get(`/flights?limit=${paginationLimit}&offset=${newOffset}`);

				setOffset(newOffset);
				setFlights([...flights, ...response?.data?.data]);

				if (paginationLimit * (newOffset + 1) > response?.data?.pagination?.total) {
					setHasFlightsRemaining(false);
				}
			} catch (error) {
				console.log('Failed to load !');
			};
		}

		fetchData();
	// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [flights, hasFlightsRemaining]);

	return (
		<FlightContext.Provider value={{
			flights,
			toggleActiveFlight,
			activeFlights
		}}>
			{ children }
		</FlightContext.Provider>
	)
}

export const FlightContext = createContext({});