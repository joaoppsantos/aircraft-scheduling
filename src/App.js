import './App.css';
import { AircraftProvider } from './contexts/AircraftContext';
import { FlightProvider } from './contexts/FlightContext';
import { NavDate, Aircrafts, Flight, Rotation } from './components';

function App() {
	const mainComponents = [<Aircrafts />, <Rotation />, <Flight />];

	return (
		<FlightProvider>
			<AircraftProvider>
				<div className="app">
					<div className="nav-date">
						<NavDate />
					</div>
					<div className="main-container">
						{mainComponents.map((component, index) => (
								<div key={index} className="element-container">
									{component}
								</div>
							)
						)}
					</div>
				</div>
			</AircraftProvider>
		</FlightProvider>
	);
}

export default App;
