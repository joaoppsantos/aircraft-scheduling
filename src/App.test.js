import * as React from 'react';
import { render } from '@testing-library/react';
import App from './App';
import { AircraftContext } from './contexts/AircraftContext';
import { FlightContext } from './contexts/FlightContext';

beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date(2022, 1, 1));
});

afterAll(() => {
    jest.useRealTimers();
});

const flightMock = {
	"id": "AS1001",
	"departuretime": 21600,
	"arrivaltime": 26100,
	"readable_departure": "06:00",
	"readable_arrival": "07:15",
	"origin": "LFSB",
	"destination": "LFMN"
};
const mockAircraftProps = {
	ident: "GABCD",
	economySeats: 186,
	base: "EGKK",
	type: "A320",
};

describe('<App />', () => {
	it('should render correctly', () => {
		const mockFlightContext = {
			activeFlights: [
				flightMock,
			],
		};
		const mockAircraftContext = {
			selectAircraft: jest.fn(),
			currentPercentage: "50",
			setPercentage: jest.fn()
		}

		const { container } = render(
			<AircraftContext.Provider value={mockAircraftContext}>
				<FlightContext.Provider value={mockFlightContext}>
					<App {...mockAircraftProps} />
				</FlightContext.Provider>
			</AircraftContext.Provider>
		);

		expect(container).toMatchSnapshot();
	});
});
