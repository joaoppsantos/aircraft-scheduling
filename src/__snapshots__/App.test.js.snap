// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render correctly 1`] = `
<div>
  <div
    class="app"
  >
    <div
      class="nav-date"
    >
      <div>
        Mon, January 31, 2022
      </div>
    </div>
    <div
      class="main-container"
    >
      <div
        class="element-container"
      >
        <div
          class="aircrafts"
        >
          Aircrafts
        </div>
      </div>
      <div
        class="element-container"
      >
        <div
          class="rotation"
        >
          Rotation
        </div>
      </div>
      <div
        class="element-container"
      >
        <div
          class="flight"
        >
          Flight
        </div>
      </div>
    </div>
    <div
      class="schedule-container"
      id="schedule_id"
    >
      <div
        class="schedule"
      >
        <div
          class="hour-schedule"
          id="hour_0"
        />
        <div
          class="hour-schedule"
          id="hour_1"
        />
        <div
          class="hour-schedule"
          id="hour_2"
        />
        <div
          class="hour-schedule"
          id="hour_3"
        />
        <div
          class="hour-schedule"
          id="hour_4"
        />
        <div
          class="hour-schedule"
          id="hour_5"
        />
        <div
          class="hour-schedule"
          id="hour_6"
        />
        <div
          class="hour-schedule"
          id="hour_7"
        />
        <div
          class="hour-schedule"
          id="hour_8"
        />
        <div
          class="hour-schedule"
          id="hour_9"
        />
        <div
          class="hour-schedule"
          id="hour_10"
        />
        <div
          class="hour-schedule"
          id="hour_11"
        />
        <div
          class="hour-schedule"
          id="hour_12"
        />
        <div
          class="hour-schedule"
          id="hour_13"
        />
        <div
          class="hour-schedule"
          id="hour_14"
        />
        <div
          class="hour-schedule"
          id="hour_15"
        />
        <div
          class="hour-schedule"
          id="hour_16"
        />
        <div
          class="hour-schedule"
          id="hour_17"
        />
        <div
          class="hour-schedule"
          id="hour_18"
        />
        <div
          class="hour-schedule"
          id="hour_19"
        />
        <div
          class="hour-schedule"
          id="hour_20"
        />
        <div
          class="hour-schedule"
          id="hour_21"
        />
        <div
          class="hour-schedule"
          id="hour_22"
        />
        <div
          class="hour-schedule"
          id="hour_23"
        />
      </div>
    </div>
  </div>
</div>
`;
